operator/pkg/ciliumidentity/metrics.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ciliumidentity
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/utils/clock"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/metrics/metric"
+)
+
+const (
+	// LabelResource indicates resources that the metrics are attributed to.
+	LabelResource = "resource"
+	// LabelPhase indicates the phases the metrics are attributed to.
+	LabelPhase = "phase"
+
+	LabelValueCID               = "cid"
+	LabelValuePod               = "pod"
+	LabelValueEnqueuedLatency   = "enqueued"
+	LabelValueProcessingLatency = "processing"
+)
+
+func NewMetrics() *Metrics {
+	return &Metrics{
+		EventCount: metric.NewCounterVec(metric.CounterOpts{
+			Namespace: metrics.CiliumOperatorNamespace,
+			Name:      "cid_controller_work_queue_event_count",
+			Help:      "Counts processed events by CID controller work queues labeled by outcome",
+		}, []string{LabelResource, metrics.LabelOutcome}),
+
+		QueueLatency: metric.NewHistogramVec(metric.HistogramOpts{
+			Namespace: metrics.CiliumOperatorNamespace,
+			Name:      "cid_controller_work_queue_latency",
+			Help:      "Duration of CID controller work queues enqueuing and processing latencies in seconds",
+			Buckets:   append(prometheus.DefBuckets, 60, 300, 900, 1800, 3600),
+		}, []string{LabelResource, LabelPhase}),
+	}
+}
+
+type Metrics struct {
+	EventCount   metric.Vec[metric.Counter]
+	QueueLatency metric.Vec[metric.Observer]
+}
+
+func (m Metrics) meterLatency(label string, enqueueTime time.Time, startTime time.Time) {
+	enqueuedLatency := startTime.Sub(enqueueTime).Seconds()
+	m.QueueLatency.WithLabelValues(label, LabelValueEnqueuedLatency).Observe(enqueuedLatency)
+
+	processingLatency := time.Since(startTime).Seconds()
```

Comment:
I think it would be better to calculate the `processingLatency` just after `qItem.Reconcile` and then pass that interval as a parameter here.
Not that it is making a big difference as of now, but it would make it safer in case of future changes between the item processing and the call of `meterLatency`.

Reply:
Moved the calculation in the controller.